### PR TITLE
fix: Don't delegate public methods overridden by a private in the decorator

### PR DESCRIPTION
**Pull request by brunohkbx**
Public methods defined on the object was being delegated to the decorator even though it was overridden by a private method.

## Testing
1. Create a simple PORO with a public method
```ruby
class Job
  def location
    Struct.new(:id) { }
  end
end
```

1. Create a decorator and override the public method with a private
```ruby
class JobDecorator < Draper::Decorator
  delegate_all

  private

  def location
    LocationDecorator.new(object.location)
  end
end
```

3. It should raise an exception when calling `.location` from the decorator
```ruby
[9] pry(main)> JobDecorator.decorate(Job.new).location
NoMethodError: private method `location' called for #<JobDecorator:0x0000559f113ba118>
```

## References
* Issue #771

### CODE REVIEW

File: spec/draper/decorator_spec.rb [hunk 1]
```diff
@@ -664,6 +664,26 @@ def hello_world
           expect{decorator.hello_world}.to raise_error NoMethodError
           expect(decorator.methods).not_to include :hello_world
         end
+
+        context 'When decorator overrides a public method defined on the object with a private' do
```
Comment (codebycliff) [inline spec/draper/decorator_spec.rb]: I believe it's preferred to not capitalize these:

```suggestion
        context 'when decorator overrides a public method defined on the object with a private' do
```

File: spec/draper/decorator_spec.rb [hunk 2]
```diff
@@ -664,6 +664,26 @@ def hello_world
...
+            Class.new(Decorator) do
+              private
+
+              def hello_world
+                'hello world'
+              end
+            end
+          end
+
+          let(:object) { Class.new{ def hello_world; end }.new }
```
Comment (codebycliff) [inline spec/draper/decorator_spec.rb]: Small spacing issue:

```suggestion
          let(:object) { Class.new { def hello_world; end }.new }
```

**Comment by brunohkbx** [comment]
@codebycliff I was thinking the same. Could it be a breaking change? :thinking: 
What if we raise a deprecated warning instead of raising an error?

**Comment by codebycliff** [comment]
@brunohkbx I think that's definitely an option. I'm also considering just bumping the major version when I release this. It's been a while since a major version has been released, so I don't think it would be that big of a deal.

**Comment by springerigor** [comment]
@brunohkbx & @codebycliff I need your attention https://github.com/drapergem/draper/pull/875. Thanks in advance 🙂

**Pull request merged**